// garbo-lib/dist/value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeValue = void 0;
const kolmafia_1 = require("kolmafia");
const libram_1 = require("libram");
function makeValue(options = {
    quick: false,
}) {
    const regularValueCache = new Map();
    const historicalValueCache = new Map();
    const inputValues = options.itemValues
        ? [...options.itemValues.entries()].map(([item, val]) => [
            item,
            () => (0, libram_1.undelay)(val),
        ])
        : [];
    const specialValueLookup = new Map([
        [
            (0, libram_1.$item) `Freddy Kruegerand`,
            currency(...(0, libram_1.$items) `bottle of Bloodweiser, electric Kool-Aid, Dreadsylvanian skeleton key`),
        ],
        [(0, libram_1.$item) `Beach Buck`, currency((0, libram_1.$item) `one-day ticket to Spring Break Beach`)],
        [
            (0, libram_1.$item) `Coinspiracy`,
            currency(...(0, libram_1.$items) `Merc Core deployment orders, karma shawarma`),
        ],
        [(0, libram_1.$item) `FunFunds™`, currency((0, libram_1.$item) `one-day ticket to Dinseylandfill`)],
        [(0, libram_1.$item) `Volcoino`, currency((0, libram_1.$item) `one-day ticket to That 70s Volcano`)],
        [
            (0, libram_1.$item) `Wal-Mart gift certificate`,
            currency((0, libram_1.$item) `one-day ticket to The Glaciest`),
        ],
        [(0, libram_1.$item) `Rubee™`, currency((0, libram_1.$item) `FantasyRealm guest pass`)],
        [(0, libram_1.$item) `Guzzlrbuck`, currency((0, libram_1.$item) `Never Don't Stop Not Striving`)],
        ...complexCandy(),
        [
            (0, libram_1.$item) `Merc Core deployment orders`,
            () => value((0, libram_1.$item) `one-day ticket to Conspiracy Island`),
        ],
        [
            (0, libram_1.$item) `free-range mushroom`,
            () => 3 *
                Math.max(value((0, libram_1.$item) `mushroom tea`) - value((0, libram_1.$item) `soda water`), value((0, libram_1.$item) `mushroom whiskey`) - value((0, libram_1.$item) `fermenting powder`), value((0, libram_1.$item) `mushroom filet`)),
        ],
        [
            (0, libram_1.$item) `little firkin`,
            () => averageValue(...(0, libram_1.$items) `martini, screwdriver, strawberry daiquiri, margarita, vodka martini, tequila sunrise, bottle of Amontillado, barrel-aged martini, barrel gun`),
        ],
        [
            (0, libram_1.$item) `normal barrel`,
            () => averageValue(...(0, libram_1.$items) `a little sump'm sump'm, pink pony, rockin' wagon, roll in the hay, slip 'n' slide, slap and tickle`),
        ],
        [
            (0, libram_1.$item) `big tun`,
            () => averageValue(...(0, libram_1.$items) `gibson, gin and tonic, mimosette, tequila sunset, vodka and tonic, zmobie`),
        ],
        [
            (0, libram_1.$item) `weathered barrel`,
            () => averageValue(...(0, libram_1.$items) `bean burrito, enchanted bean burrito, jumping bean burrito`),
        ],
        [
            (0, libram_1.$item) `dusty barrel`,
            () => averageValue(...(0, libram_1.$items) `spicy bean burrito, spicy enchanted bean burrito, spicy jumping bean burrito`),
        ],
        [
            (0, libram_1.$item) `disintegrating barrel`,
            () => averageValue(...(0, libram_1.$items) `insanely spicy bean burrito, insanely spicy enchanted bean burrito, insanely spicy jumping bean burrito`),
        ],
        [
            (0, libram_1.$item) `moist barrel`,
            () => averageValue(...(0, libram_1.$items) `cast, concentrated magicalness pill, enchanted barbell, giant moxie weed, Mountain Stream soda`),
        ],
        [
            (0, libram_1.$item) `rotting barrel`,
            () => averageValue(...(0, libram_1.$items) `Doc Galaktik's Ailment Ointment, extra-strength strongness elixir, jug-o-magicalness, Marquis de Poivre soda, suntan lotion of moxiousness`),
        ],
        [
            (0, libram_1.$item) `mouldering barrel`,
            () => averageValue(...(0, libram_1.$items) `creepy ginger ale, haunted battery, scroll of drastic healing, synthetic marrow, the funk`),
        ],
        [
            (0, libram_1.$item) `barnacled barrel`,
            () => averageValue(...(0, libram_1.$items) `Alewife™ Ale, bazookafish bubble gum, beefy fish meat, eel battery, glistening fish meat, ink bladder, pufferfish spine, shark cartilage, slick fish meat, slug of rum, slug of shochu, slug of vodka, temporary teardrop tattoo`),
        ],
        [
            (0, libram_1.$item) `psychoanalytic jar`,
            () => 
            // Exclude jick because he's rate-limited
            Math.max(...(0, libram_1.$items) `jar of psychoses (The Meatsmith), jar of psychoses (The Captain of the Gourd), jar of psychoses (The Crackpot Mystic), jar of psychoses (The Pretentious Artist), jar of psychoses (The Old Man), jar of psychoses (The Suspicious-Looking Guy)`.map((jar) => value(jar))),
        ],
        [
            (0, libram_1.$item) `warbear whosit`,
            () => (0.35 *
                averageValue(...(0, libram_1.$items) `warbear auto-anvil, warbear chemistry lab, warbear high-efficiency still, warbear induction oven, warbear jackhammer drill press, warbear LP-ROM burner, warbear energy bracers, warbear exhaust manifold, warbear exo-arm, warbear foil hat, warbear laser beacon, warbear oil pan`) +
                0.65 *
                    averageValue(...(0, libram_1.$items) `warbear metalworking primer, warbear beeping telegram, warbear gyrocopter, warbear procedural hilarity drone, warbear robo-camouflage unit, warbear sequential gaiety distribution system`)) /
                100,
        ],
        ...(0, libram_1.$items) `worthless gewgaw, worthless knick-knack, worthless trinket`.map((i) => [
            i,
            currency(...(0, libram_1.$items) `seal tooth, chisel, petrified noodles, jabañero pepper, banjo strings, hot buttered roll, wooden figurine, ketchup, catsup, volleyball`, ...((0, kolmafia_1.myClass)() === (0, libram_1.$class) `Seal Clubber`
                ? (0, libram_1.$items) `figurine of an ancient seal`
                : [])),
        ]),
        [
            (0, libram_1.$item) `Boris's key`,
            () => value((0, libram_1.$item) `Boris's key lime`) - value((0, libram_1.$item) `lime`),
        ],
        [
            (0, libram_1.$item) `Jarlsberg's key`,
            () => value((0, libram_1.$item) `Jarlsberg's key lime`) - value((0, libram_1.$item) `lime`),
        ],
        [
            (0, libram_1.$item) `Sneaky Pete's key`,
            () => value((0, libram_1.$item) `Sneaky Pete's key lime`) - value((0, libram_1.$item) `lime`),
        ],
        [
            (0, libram_1.$item) `fat loot token`,
            currency(...(0, libram_1.$items) `Boris's key, Jarlsberg's key, Sneaky Pete's key, Boris's ring, Jarlsberg's earring, Sneaky Pete's breath spray, potato sprout, sewing kit, Spellbook: Singer's Faithful Ocelot, Spellbook: Drescher's Annoying Noise, Spellbook: Walberg's Dim Bulb, dried gelatinous cube`),
        ],
        ...inputValues,
    ]);
    const exclusions = new Set([
        // For tradeable items which can be "consumed" infinitely
        (0, libram_1.$item) `ChibiBuddy™ (off)`,
    ]);
    function saleValue(item, useHistorical) {
        if (useHistorical) {
            if ((0, kolmafia_1.historicalAge)(item) <= 7.0 && (0, kolmafia_1.historicalPrice)(item) > 0) {
                const isMallMin = (0, kolmafia_1.historicalPrice)(item) === Math.max(100, 2 * (0, kolmafia_1.autosellPrice)(item));
                return isMallMin ? (0, kolmafia_1.autosellPrice)(item) : 0.9 * (0, kolmafia_1.historicalPrice)(item);
            }
        }
        return (0, libram_1.getSaleValue)(item);
    }
    function currency(...items) {
        const unitCost = items.map((i) => {
            const coinmaster = kolmafia_1.Coinmaster.all().find((c) => (0, kolmafia_1.sellPrice)(c, i) > 0);
            if (!coinmaster) {
                throw `Invalid coinmaster item ${i}`;
            }
            else {
                return [i, (0, kolmafia_1.sellPrice)(coinmaster, i)];
            }
        });
        return () => Math.max(...unitCost.map(([item, cost]) => value(item) / cost));
    }
    function complexCandy() {
        const candies = kolmafia_1.Item.all().filter((i) => i.candyType === "complex");
        const candyLookup = [[], [], [], [], []];
        for (const candy of candies) {
            const id = (0, kolmafia_1.toInt)(candy) % 5;
            if (candy.tradeable) {
                candyLookup[id].push(candy);
            }
        }
        const candyIdPrices = candies
            .filter((i) => !i.tradeable)
            .map((i) => [
            i,
            () => Math.min(...candyLookup[(0, kolmafia_1.toInt)(i) % 5].map((i) => value(i))),
        ]);
        return candyIdPrices;
    }
    function value(inputItem, useHistorical = false) {
        const { item, quantity } = inputItem instanceof kolmafia_1.Item ? { item: inputItem, quantity: 1 } : inputItem;
        if (exclusions.has(item))
            return 0;
        useHistorical ||= options.quick;
        const cachedValue = regularValueCache.get(item) ??
            (useHistorical ? historicalValueCache.get(item) : undefined);
        if (cachedValue === undefined) {
            const specialValueCompute = specialValueLookup.get(item);
            const value = specialValueCompute?.() ?? saleValue(item, useHistorical);
            (useHistorical ? historicalValueCache : regularValueCache).set(item, value);
            return value;
        }
        return quantity * cachedValue;
    }
    function averageValue(...items) {
        return (0, libram_1.sum)(items, value) / items.length;
    }
    return {
        averageValue,
        value,
    };
}
exports.makeValue = makeValue;
