// garbo-lib/dist/wanderer/eightbit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.eightbitFactory = exports.TREASURE_HOUSE_FAT_LOOT_TOKEN_COST = exports.bonusColor = void 0;
const kolmafia_1 = require("kolmafia");
const lib_1 = require("./lib");
const libram_1 = require("libram");
exports.bonusColor = ["black", "blue", "green", "red"];
exports.TREASURE_HOUSE_FAT_LOOT_TOKEN_COST = 20000;
// taken from tour guide, who took it from beldur
const locationColor = {
    black: (0, libram_1.$location) `Vanya's Castle`,
    blue: (0, libram_1.$location) `Megalo-City`,
    green: (0, libram_1.$location) `Hero's Field`,
    red: (0, libram_1.$location) `The Fungus Plains`,
};
function value(color, options) {
    const denominator = options.ascend
        ? (0, libram_1.get)("8BitScore") - exports.TREASURE_HOUSE_FAT_LOOT_TOKEN_COST
        : exports.TREASURE_HOUSE_FAT_LOOT_TOKEN_COST;
    return ((options.itemValue((0, libram_1.$item) `fat loot token`) *
        (0, kolmafia_1.eightBitPoints)(locationColor[color])) /
        denominator);
}
function eightbitFactory(type, locationSkiplist, options) {
    if ((0, libram_1.have)((0, libram_1.$item) `continuum transfunctioner`) &&
        type !== "backup" &&
        (0, libram_1.get)("8BitScore") < exports.TREASURE_HOUSE_FAT_LOOT_TOKEN_COST) {
        return exports.bonusColor
            .map((color) => new lib_1.WandererTarget(`8-bit (${color})`, locationColor[color], value(color, options)))
            .filter((t) => !locationSkiplist.includes(t.location));
    }
    return [];
}
exports.eightbitFactory = eightbitFactory;
